rs/ledger_suite/icp/ledger/src/main.rs

```diff
@@ -801,41 +808,102 @@ fn post_upgrade(args: Option<LedgerCanisterPayload>) {
                 };
             ledger_state
         });
-    }
 
-    if ledger.ledger_version > LEDGER_VERSION {
-        panic!(
-            "Trying to downgrade from incompatible version {}. Current version is {}.",
-            ledger.ledger_version, LEDGER_VERSION
-        );
-    }
-    ledger.ledger_version = LEDGER_VERSION;
+        let upgrade_from_version = ledger.ledger_version;
+        if ledger.ledger_version > LEDGER_VERSION {
+            panic!(
+                "Trying to downgrade from incompatible version {}. Current version is {}.",
+                ledger.ledger_version, LEDGER_VERSION
+            );
+        }
+        ledger.ledger_version = LEDGER_VERSION;
 
-    if let Some(args) = args {
-        match args {
+        if let Some(args) = args {
+            match args {
             LedgerCanisterPayload::Init(_) => trap_with("Cannot upgrade the canister with an Init argument. Please provide an Upgrade argument."),
             LedgerCanisterPayload::Upgrade(upgrade_args) => {
                 if let Some(upgrade_args) = upgrade_args {
                     ledger.upgrade(upgrade_args);
                 }
         }
     }
+        }
+        set_certified_data(
+            &ledger
+                .blockchain
+                .last_hash
+                .map(|h| h.into_bytes())
+                .unwrap_or([0u8; 32]),
+        );
+        PRE_UPGRADE_INSTRUCTIONS_CONSUMED
+            .with(|n| *n.borrow_mut() = pre_upgrade_instructions_consumed);
+
+        if upgrade_from_version == 0 {
+            set_ledger_state(LedgerState::Migrating(LedgerField::Allowances));
+            print("Upgrading from version 0 which does not use stable structures, clearing stable allowance data.");
+            clear_stable_allowance_data();
+            ledger.clear_arrivals();
+        }
+    }
+    if !is_ready() {
+        print("Migration started.");
+        migrate_next_part(
+            MAX_INSTRUCTIONS_PER_UPGRADE.saturating_sub(pre_upgrade_instructions_consumed),
```

Comment:
Here we're ignoring post-upgrade instructions already spent.

Reply:
We take already spent instructions when doing the check `while instruction_counter() < instruction_limit {` - we check whether all instructions used in post_upgrade don't exceed the above limit.